code/game/turfs/turf.dm

```diff
@@ -132,40 +132,38 @@
 	stack_trace("Non movable passed to turf CanPass : [mover]")
 	return FALSE
 
-/turf/Enter(atom/movable/mover as mob|obj, atom/forget as mob|obj|turf|area)
+//returns which obstruction would prevent entering. Update if /turf/enter is updated.
+/turf/proc/CheckEnteringObstruction(atom/movable/mover, atom/forget)
 	// First, make sure it can leave its square
 	if(isturf(mover.loc))
 		// Nothing but border objects stop you from leaving a tile, only one loop is needed
 		for(var/obj/obstacle in mover.loc)
 			if(!obstacle.CheckExit(mover, src) && obstacle != mover && obstacle != forget)
-				mover.Collide(obstacle)
-				return FALSE
-
+				return obstacle
 	//Then, check the turf itself
 	if (!src.CanPass(mover, src))
-		mover.Collide(src)
-		return FALSE
-
-
+		return src
 	var/atom/movable/topmost_bump
 	var/top_layer = FALSE
-
 	//Next, check objects to block entry that are on the border
 	for(var/atom/movable/obstacle in src)
 		if(!obstacle.CanPass(mover, mover.loc, 1) && (forget != obstacle))
 			if(obstacle.flags_1 & ON_BORDER_1)
-				mover.Collide(obstacle)
-				return FALSE
+				return obstacle
 			else
 				if(obstacle.layer > top_layer)
 					topmost_bump = obstacle
 					top_layer = obstacle.layer
-
 	if(topmost_bump)
-		mover.Collide(topmost_bump)
-		return FALSE
+		return topmost_bump
 
-	return TRUE //Nothing found to block so return success!
+/turf/Enter(atom/movable/mover as mob|obj, atom/forget as mob|obj|turf|area)
```

Comment:
I'm not sure adding procall on top of one of the busiest functions is good idea here.

Reply:
You could potentially make it a define/macro

But this is so he doesn't duplicate code, so I think it's worth the tradeoff

kev do you think you could make the collide check a define?